grammar/generics.g

```diff
@@ -0,0 +1,35 @@
+Generics = "<" params:GenericParam* % "," ","? ">";
+GenericParam = attrs:OuterAttr* kind:GenericParamKind;
+GenericParamKind =
+    Lifetime:{ name:LIFETIME { ":" bounds:LifetimeBound* % "+" "+"? }? } |
+    Type:{ name:IDENT { ":" bounds:TypeBound* % "+" "+"? }? { "=" default:Type }? };
+
+ForAllBinder = "for" generics:Generics;
+
+WhereClause = "where" bounds:WhereBound* % "," ","?;
+WhereBound =
+    Lifetime:{ lt:LIFETIME ":" bounds:LifetimeBound* % "+" "+"? } |
+    Type:{ binder:ForAllBinder? ty:Type ":" bounds:TypeBound* % "+" "+"? } |
```

Comment:
Disturbing fact: the following is valid Rust syntax:

```
trait X {}

fn foo<T>()
where for<'a> for<'b> X: Clone
{}
```

however, syntactically these `for`'s are different: the first one is a part of where predicate, the second one is a part of the type. Our ambiguity testing should ideally expose this problem.

Reply:
Yupp, we basically have a precedence rule that eagerly takes the first encountered `for<'a>` and uses it for the whole `T: A + B +...`.

Even more fun is the optional *but not nestable* parens around a bound.
@Centril suggested we should make it a proper expression grammar with parens and `+` as a binop between bounds, which I feel pretty sympathetic towards.

`where i32: (for<> Copy)` looks weird but is perfectly valid today!

Reply:
> Why?

Well, it would explain the parens. And they do feel maybe like entities that can be arbitrarily composed.